### PR TITLE
feat(nim): add support for nimble project package version

### LINKED ISSUE #2454: Package version: Nimble project support
#### Describe the solution you'd like

Nimble projects contain a `version` property that should be included in the package version module.

#### Implementing

When a file with the extension `.nimble` is in the current directory the version can be read by using [`nimble dump`](https://github.com/nim-lang/nimble#nimble-dump) to parse & evaluate the file. The output can be in INI-compatible format (default) or JSON with the `--json` flag. You could use a parser for INI or JSON, or just use a regex (which is probably faster than parsing).

Running the command `nimble dump` in the project will either output the project information or error with an non-zero exit code.

#### Test cases

In a new directory create a file containing `example.nimble` filled with

```nims
# Comments
version     = "1.2.3"
author      = "Example Author"
description = "A Nim package"
license     = "MIT"
```

Running `nimble dump` should output

```ini
name: "tmp"
version: "1.2.3"
author: "Example Author"
desc: "A Nim package"
license: "MIT"
skipDirs: ""
skipFiles: ""
skipExt: ""
installDirs: ""
installFiles: ""
installExt: ""
requires: ""
bin: ""
binDir: ""
srcDir: ""
backend: "c"
```

and with `--json`

```json
{
  "name": "tmp",
  "version": "1.2.3",
  "author": "Example Author",
  "desc": "A Nim package",
  "license": "MIT",
  "skipDirs": [],
  "skipFiles": [],
  "skipExt": [],
  "installDirs": [],
  "installFiles": [],
  "installExt": [],
  "requires": [],
  "bin": [],
  "binDir": "",
  "srcDir": "",
  "backend": "c"
}
```

### PR DESCRIPTION
**Pull request by aeruhxi**
This PR adds support for the package module for a nimble project to show version number in the project directory.

#### Description

~~I have added a new function `find_file`  that takes a directory and a predicate. It iterates over all files in a given directory and returns a filename if it satisfies the given predicate. This function is used to detect if a given directory is a nimble project by checking `file_name.ends_with(".nimble")`.~~

If a directory is found to be a nimble project, `extract_nimble_version` then calls `nimble dump --json` to get the package version. I used `json` instead of the default `ini` format for dump because I could not find an existing `ini` parser used in the project but `json` parser was already there used predominantly for most packages.

##### Update
It now uses context to check for the existence of `.nimble` file.

##### Alternative implementation?
After already writing `find_file`, I later found out that Instead of using `find_file`, I could directly call `nimble dump` and checks its exit code, because it fails if it's not a nimble project, to determine a nimble project. I want to have others' opinion on this if that would be a better implementation idea.

#### Motivation and Context
The motivation and context is in the linked issue.

Closes #2454.

#### Screenshots (if appropriate):
image

#### How Has This Been Tested?

- [ ] I have tested using **MacOS** (I don't have an access to it)
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows** (I don't have an access to it)

#### Checklist:

This PR is a WIP. Docs and tests are yet to be done.
- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

### CODE REVIEW

File: src/modules/package.rs [hunk 1]
```diff
@@ -175,6 +186,8 @@ fn extract_meson_version(file_contents: &str) -> Option<String> {
 fn get_package_version(base_dir: &Path, config: &PackageConfig) -> Option<String> {
     if let Ok(cargo_toml) = utils::read_file(base_dir.join("Cargo.toml")) {
         extract_cargo_version(&cargo_toml)
+    } else if utils::find_file(base_dir, |file_name| file_name.ends_with(".nimble")).is_some() {
```
Comment (davidkna) [inline src/modules/package.rs:189]: Can you have this at least make the initial check for `nimble` files via `context`, like the language modules do? It keeps a list of files and file extensions.
```rust
context
  .try_begin_scan()?
  .set_extensions(&["nimble"])
  .is_match();
```

File: src/modules/package.rs [hunk 2]
```diff
@@ -54,6 +53,16 @@ fn extract_cargo_version(file_contents: &str) -> Option<String> {
     Some(formatted_version)
 }
 
+fn extract_nimble_version() -> Option<String> {
+    let cmd_output = utils::exec_cmd("nimble", &["dump", "--json"], Duration::from_millis(500))?;
```
Comment (davidkna) [inline src/modules/package.rs:57]: Can you change this to use `context.exec_cmd` so this uses the global timeout?

**Review by davidkna — CHANGES_REQUESTED** [review]
I think the current approach and not always calling `nimble dump` is better for general performance.

File: src/modules/package.rs (hunk 2 above)
Comment (aeruhxi) [reply src/modules/package.rs:57]: Yeah, I've already changed it. Sorry about that. I was trying to push it along with tests. I will work on it sometime later.

**Comment by aeruhxi** [comment]
It has been complete from my side now. I don't own a dev macos or windows machine.

File: docs/config/README.md [hunk 3]
```diff
@@ -1939,11 +1939,12 @@ symbol = "☁️ "
 ## Package Version
 
 The `package` module is shown when the current directory is the repository for a
-package, and shows its current version. The module currently supports `npm`, `cargo`,
+package, and shows its current version. The module currently supports `npm`, `nimble`, `cargo`,
 `poetry`, `composer`, `gradle`, `julia`, `mix` and `helm` packages.
 
 - **npm** – The `npm` package version is extracted from the `package.json` present
   in the current directory
+- **nimble** - The `nimble` package version is extracted from the `nimble dump` present
```
Comment (davidkna) [inline docs/config/README.md:1947]: Can you reword this something like this:

```suggestion
- **nimble** - The `nimble` package version is extracted from the `*.nimble` file present in the current directory with the `nimble dump` command
```

File: src/modules/package.rs [hunk 4]
```diff
@@ -248,6 +264,86 @@ mod tests {
...
+        let expected = Some(format!(
+            "is {} ",
+            Color::Fixed(208).bold().paint(format!("📦 {}", "v0.1.0"))
+        ));
+
+        assert_eq!(actual, expected);
+        project_dir.close()
+    }
+
+    #[test]
```
Comment (davidkna) [inline src/modules/package.rs:326]: Another test with a nimble file but no nimble command available would be nice.

**Comment by davidkna** [comment]
> I don't own a dev macos or windows machine.

That's fine.

File: src/modules/package.rs (hunk 4 above)
Comment (aeruhxi) [reply src/modules/package.rs:326]: Sure. I've just added it.

**Comment by vladimyr** [comment]
Let me try something before we merge this, looks good otherwise 👍

**Comment by vladimyr** [comment]
> Let me try something before we merge this, looks good otherwise

Ignore me, my attempts failed so let's continue reviewing it as is.

**Comment by aeruhxi** [comment]
Should I rebase and resolve conflicts?

**Comment by vladimyr** [comment]
> Should I rebase and resolve conflicts?

Please do 🚀

**Comment by aeruhxi** [comment]
> > Should I rebase and resolve conflicts?
> 
> Please do

Done.

**Review by davidkna — APPROVED** [review]
LGTM

**Comment by andytom** [comment]
Sorry @aeruhxi can you fix the merge conflicts again and then we can get this merged.

**Comment by aeruhxi** [comment]
> Sorry @aeruhxi can you fix the merge conflicts again and then we can get this merged.

Sure, no problem.

**Comment by andytom** [comment]
Thank you for your contribution @aeruhxi and thank you for the reviews @davidkna and @vladimyr.

**Pull request merged**